drivers: serial: define default values for basic options

yishai1999 opened the pull request:
Defined default values for baudrate, parity, stop bits, and data bits. This removes complexity and obfuscation from the code.

I made the changes only to a few drivers to gauge interest - If this is accepted I will do the rest too.

yishai1999 committed:
drivers: serial: define default values for basic options

Defined default values for baudrate, parity, stop bits, and data bits.
This removes complexity and obfuscation from the code.

Signed-off-by: Yishai Jaffe <yishai1999@gmail.com>

nordicjm commented on dts/bindings/serial/uart-controller.yaml:
```diff
@@ -10,12 +10,14 @@ properties:
     description: Clock frequency information for UART operation
   current-speed:
     type: int
+    default: 115200
```
Lacking justification, see https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

yishai1999 replied:
Do you think that's true for all of the properties or just baudrate?
Also, Shouldn't it be a required property at least?

nordicjm replied:
All of them. If you add a default then they are required as the default will always be present unless something overrides

jhedberg replied:
What I was trying to propose in https://github.com/zephyrproject-rtos/zephyr/pull/83312#discussion_r1895600416 was to do this in the per-driver binding rather than the base binding.

As pointed out by @yishai1999 in https://github.com/zephyrproject-rtos/zephyr/pull/83312#discussion_r1895542662 there seems to be several drivers using `DT_INST_*_OR()` to have defaults in the driver code itself for these properties, e.g. here:
https://github.com/zephyrproject-rtos/zephyr/blob/6dce23326501a0850d2af9ae2af791ec0e7b6037/drivers/serial/uart_esp32.c#L1058-L1062

If there really *has* to be a default in these cases (which I'm not convinced of) having per-driver DT binding default values seems like the lesser of two evils, since then you can at least see the actual value that was used by looking at `build/zephyr/zephyr.dts`. However, either way, it'd be good to establish some consistent approach for this across all UART drivers.

yishai1999 replied:
@nordicjm What do you think about it now?

nordicjm replied:
Where's the required description as per the above link stating where the default value comes from? E.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/sensor/vishay%2Cvcnl36825t.yaml#L47

yishai1999 replied:
I guess we would have to ask the people who added the defaults in the drivers' code (@knthm , @bwitherspoon , @gohshunjing  + others for more drivers) but that fact is - They have defaults, they're just not written in the DTS which as @jhedberg said, makes it less easy to understand.

jerome-pouiller replied:
Indeed, any sane driver will use 115200 8|N|1 for default values. We can probably use these values in `uart-controller.yaml`.

yishai1999 replied:
> I also agree this is a better solution than current `DT_INST_*_OR()`. Regarding the rationale behind these defaults, I guess we expect all Zephyr targets to use same default (otherwise it would be a mess), so it should be documented somewhere and we should be able to refer to that, right ?

Sorry but which solution is the "this" that you are referring to?

bwitherspoon replied:
It has been sometime since I implemented the TI driver, but I am pretty sure I followed the other drivers which defaulted to 115200 8/N/1. As a user I would be very surprised if the default values were not this.

+1 This is a good change.

erwango replied:
> Sorry but which solution is the "this" that you are referring to?

Proposed use of binding defaults.

rruuaanng commented on dts/bindings/serial/ti,cc13xx-cc26xx-uart.yaml:
```diff
@@ -13,3 +13,15 @@ properties:
+    default: "none"
+
+  stop-bits:
+    default: "1"
```
Why do you using string not a number. Like `data-bits`.
(Sorry, I'm mobile now, so don't see other file.)

yishai1999 replied:
That's how it's defined in the enum. Probably because there are non integer values too (0.5 and 1.5)

rruuaanng replied:
Oh yeah, I seem to have forgotten, this is the general enumeration value of the serial.

rruuaanng commented on dts/bindings/serial/st,stm32-uart-base.yaml:
```diff
@@ -25,6 +25,18 @@ properties:
+    default: "none"
+
+  stop-bits:
+    default: "1"
```
+1

nashif commented:
@nordicjm  please revisit, your comment was about a file that is not part of this patchest anymore.

nordicjm commented:
> @nordicjm please revisit, your comment was about a file that is not part of this patchest anymore.

Same thing applies, https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values clearly states a description field is required for when a default is used, so said description needs adding to these fields

jhedberg commented:
> Same thing applies, https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values clearly states a description field is required for when a default is used, so said description needs adding to these fields

Note that this PR isn't adding any new defaults, rather just moving them from the C code (less discoverable) into the bindings (more discoverable). That said, documenting them may indeed be justified.

yishai1999 commented:
> > @nordicjm please revisit, your comment was about a file that is not part of this patchest anymore.
> 
> Same thing applies, https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values clearly states a description field is required for when a default is used, so said description needs adding to these fields

I added a description. The `parity` property in `uart-controller.yaml` already has a default in the description so I just added the `default` attribute to it.

jhedberg commented on dts/bindings/serial/intel,lw_uart.yaml:
```diff
@@ -17,3 +17,12 @@ properties:
     description: |
       Baud rate cannot be changed by software (Divisor register is not writable)
+
+  parity:
+    default: "none"
+
+  stop-bits:
+    default: "1"
+
+  data-bits:
+    default: 8
```
Seems like you forgot to add the description here?

yishai1999 replied:
My bad, fixed.

yishai1999 commented:
Applied said changes on the rest of the drivers

jerome-pouiller commented:
This PR is an improvement. However, I believe we should go further and define these defaults in 
`uart-controller.yaml`.
